File: lingting-aws/src/main/java/live/lingting/framework/aws/s3/AwsS3Utils.java
package live.lingting.framework.aws.s3;

import live.lingting.framework.time.DatePattern;
import lombok.experimental.UtilityClass;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

/**
 * @author lingting 2024-09-19 15:20
 */
@UtilityClass
public class AwsS3Utils {

	/**
	 * 10M
	 */
	public static final long MULTIPART_DEFAULT_PART_SIZE = 10485760;

	/**
	 * 5G
	 */
	public static final long MULTIPART_MAX_PART_SIZE = 5368709120L;

	/**
	 * 100K
	 */
	public static final long MULTIPART_MIN_PART_SIZE = 102400;

	public static final long MULTIPART_MAX_PART_COUNT = 1000;

	public static final String PAYLOAD_UNSIGNED = "UNSIGNED-PAYLOAD";

	public static final String HEADER_PREFIX = "x-amz";

	public static final String HEADER_DATE = HEADER_PREFIX + "-date";

	public static final String HEADER_CONTENT_SHA256 = HEADER_PREFIX + "-content-sha256";

	public static final String HEADER_TOKEN = HEADER_PREFIX + "-security-token";

	public static final String HEADER_ACL = HEADER_PREFIX + "-acl";

	public static String format(LocalDateTime dateTime, DateTimeFormatter formatter) {
		ZonedDateTime atZone = dateTime.atZone(DatePattern.SYSTEM_ZONE_ID);
		ZonedDateTime atGmt = atZone.withZoneSameInstant(DatePattern.GMT_ZONE_ID);
		return formatter.format(atGmt);
	}
	public static LocalDateTime parse(String string, DateTimeFormatter formatter) {
		LocalDateTime source = LocalDateTime.parse(string, formatter);
		ZonedDateTime atGmt = source.atZone(DatePattern.GMT_ZONE_ID);
		ZonedDateTime atZone = atGmt.withZoneSameInstant(DatePattern.SYSTEM_ZONE_ID);
		return atZone.toLocalDateTime();
	}

	public static String encode(String s) {
		return URLEncoder.encode(s, StandardCharsets.UTF_8);
	}

}

File: lingting-aws/src/main/kotlin/live/lingting/framework/aws/s3/AwsS3Utils.kt
package live.lingting.framework.aws.s3

import live.lingting.framework.data.DataSize
import live.lingting.framework.time.DatePattern
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * @author lingting 2024-09-19 15:20
 */
object AwsS3Utils {
    /**
     * 10M
     */
    val MULTIPART_DEFAULT_PART_SIZE = DataSize.ofMb(10)

    /**
     * 5G
     */
    val MULTIPART_MAX_PART_SIZE = DataSize.ofGb(5)

    /**
     * 100K
     */
    val MULTIPART_MIN_PART_SIZE = DataSize.ofKb(100)

    const val MULTIPART_MAX_PART_COUNT = 1000L

    const val PAYLOAD_UNSIGNED: String = "UNSIGNED-PAYLOAD"

    const val HEADER_PREFIX: String = "x-amz"

    const val HEADER_DATE: String = "$HEADER_PREFIX-date"

    const val HEADER_CONTENT_SHA256: String = "$HEADER_PREFIX-content-sha256"

    const val HEADER_TOKEN: String = "$HEADER_PREFIX-security-token"

    const val HEADER_ACL: String = "$HEADER_PREFIX-acl"

    const val HEADER_PREFIX_META: String = "$HEADER_PREFIX-meta-"

    @JvmStatic
    fun format(dateTime: LocalDateTime, formatter: DateTimeFormatter): String {
        val atZone = dateTime.atZone(DatePattern.SYSTEM_ZONE_ID)
        val atGmt = atZone.withZoneSameInstant(DatePattern.GMT_ZONE_ID)
        return formatter.format(atGmt)
    }

    @JvmStatic
    fun parse(string: String, formatter: DateTimeFormatter): LocalDateTime {
        val source = LocalDateTime.parse(string, formatter)
        val atGmt = source.atZone(DatePattern.GMT_ZONE_ID)
        val atZone = atGmt.withZoneSameInstant(DatePattern.SYSTEM_ZONE_ID)
        return atZone.toLocalDateTime()
    }

    @JvmStatic
    fun encode(s: String): String {
        return URLEncoder.encode(s, StandardCharsets.UTF_8)
    }
}
